rfcs/20220309-feature_flags.md

```diff
@@ -0,0 +1,219 @@
+
+  Everyone wants to build exciting features, not exercise Git-fu when there are conflicts or when it's time to finally merge the feature in.
+
+## Goals
+
+Feature flags, as an alternative to the current methodology, should allow us to do the following:
+
+### Must haves
+
+These features are strict requirements and solutions not providing them will be disqualified.
+
+1. Develop right on `main`, keeping visibility hidden until we can go live.
+   1. Allow code for both live and in-dev features to coexist.
+   1. Deployments to prod such as fixes should not be blocked.
+1. Keep the code separate from the feature visibility configuration[^separate].
+1. Toggle the visibility of features based on the environment.
+
+[^separate]: Separate not in the sense of being placed somewhere else, just being untangled or decoupled.
+
+### Good to haves
+
+These features are not strictly needed and a solution lacking them will not be discarded.
+
+1. Enable features to go-live without technically needing a new release.
+1. Associate certain metadata or sub-attributes linked to features such as conditions for toggling etc.
+1. Nested feature flags, with the parent flag acting as shortcut to toggle all its children.
+
+## Implementation
+
+This RFC considers feature flags for the frontend. The API can just use `beta/` instead of `v1/` to sufficiently mark unstable endpoints.
+
+### States
+
+Each flag can have 3 states.
+
+- **enabled** (`on`)
+
+  There is no fencing, everyone can access the feature, even if they don't like or want it.
+
+- **switchable** (`any`)
+
+  The feature is fenced by default but a user can opt-in. Preferences may be recorded.
+
+- **disabled** (`off`)
+
+  The feature is fenced and inaccessible, at least not visible. What's ever hidden in open-source?
+
+### Sources
+
+Flags must first be defined somewhere from where they can be loaded into the application.
+
+#### 1. Environment variables
+
+##### Pros
+
+1. Effortless to set up.
+1. Universal way to configure applications.
+1. Low maintenance as every feature can be one env var.
+1. Fairly easy to test considering env vars can be set per-session, per-command, per-container and per-machine (as needed).
+1. Managing a couple of `.env` files is not hard.
+1. I'm not entirely sure but server-side envs could help tree-shake the off features in prod.
+
+##### Cons
+
+1. Will need a deployment to change the environment variables.
+1. Can get complicated if we need to build a lot of scaffolding around it.
+1. Managing a lot of `.env` files is not easy.
+1. Flags will be primitive and will not allow complex data.
+1. Nested flags and flags with metadata will need hacks.
+
+##### Considerations
+
+Feature flag env vars can be prefixed with a string like `FF_` to separate them from the other env vars.
+
+The limitation for the data to be a string means any extra metadata must be JSON-stringified and then parsed in the code. Only primitive flag states can be handled conveniently.
```

Comment:
I didn't have any specific metadata in mind but I wanted to try to support that because we might need it in the future, especially when running experiments or A/B tests, [like how Growthbook does it](https://docs.growthbook.io/app/features).